File: baekjoon_all/10000+/boj_10974.cpp
// Solve 2023-01-15
// Update 2024-02-12

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int n;
int seq[8];

void recur(int depth = 0, int selected = 0) {
    if (depth == n) {
        for (int i = 0; i < n; i++) {
            cout << seq[i] << ' ';
        }

        cout << '\n';
        return;
    }

    for (int i = 1; i <= n; i++) {
        if ((selected & 1 << i) == 0) {
            seq[depth] = i;
            recur(depth + 1, selected | 1 << i);
        }
    }
}

int main() {
    FASTIO;

    cin >> n;

    recur();

    return 0;
}

File: baekjoon_all/10000+/boj_10974.java
// Solve 2024-02-12

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class boj_10974 {

    static StringBuilder sb;
    static int n;
    static int[] seq;

    static void recur(int depth, int selected) {
        if (depth == n) {
            for (int x : seq) {
                sb.append(x).append(" ");
            }

            sb.append("\n");
            return;
        }

        for (int i = 1; i <= n; i++) {
            if ((selected & 1 << i) == 0) {
                seq[depth] = i;
                recur(depth + 1, selected | 1 << i);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        n = Integer.parseInt(br.readLine());
        seq = new int[n];
        recur(0, 0);

        System.out.print(sb);
        br.close();
    }

}
